2-explicit-polymorphism/2-explicit-polymorphism.md

```diff
@@ -0,0 +1,147 @@
+a System F with holes than a Simply Typed Lambda Calculus with holes.
+
+# Type Language
+
+We'd have to extend UHTyp.t with `TyVar` and `Forall` cases:
+
+```reason
+/* UHTyp.re */
+
+type operand =
+...
+| Forall(TPat.t, t);
+| TyVar(VarErrStatus.t, Var.t)
+  ```
+
+  A type pattern is the part of the language between the `forall` and the `.`
+  in `forall a. <type>`. Type patterns will be very simple to begin with:
+
+  ```reason
+  /* TPat.t */
+  type t =
+  | Hole(MetaVar.t)
+| Var(Var.t)
+  ```
+
+  <!-- TODO: Talk about how TPat.t's might expand in the future -->
+
+# Expression Language
+
+  ```reason
+  /* UHExp.t */
+  type operand =
+  ...
+  | TyLam(TPat.t, block)
+| TyVar(TPat.t)
+  ```
+
+# Construct actions
+
+  We'll be adding the following to `Construct(shape)`:
+
+  ```
+  type shape =
+  ...
+  | SForall
+  | SType
+  ```
+
+  `Construct(SForall)` will be triggered when the user types "forall" followed by
+  a space in a type hole. This will add a Forall type with the cursor at the type
+  pattern hole and a hole as the body.
+
+  `Construct(SType)` will be triggered when the user types "type" followed by a
+  space in an expression hole. This will add a `TyVar.t` in the expression with a
+  type pattern hole.
+
+# Concrete Syntax / Representation
+
+  Forall's appear within types and look like: `forall a. ?`, where `a` is
+  a TyVar.t.
+
+  There are Type Vars types and expressions. Type Vars in types appear as a
+  string of lowercase letter without any prefix or constructor. Type Vars in
+  expressions appear as `(type a)` where `a` is also a string of lowercase
+  letters.
+
+  Type Lambdas are written like a lambda: you start with `\`. Then you write
+  "type" followed by a space which will construct a type lambda. If however, you
+  type `type<space>` into an existing lambda that has a non-empty type
+  annotation, the pattern of the lambda will be put into an error state until
+  either you delete the space following "type" or the type annotation. The final
+  representation is `λ(type a).{ }.
+
+# Backspace actions
+
+  What happens when you press backspace at the following positions? (`|`
+  represents the cursor.) I don't fully understand how this works visual
+  indicators. Any backspace that does not change the tree or the cursor
+  but only changes the visual indicator is not taken into account here.
+
+## Forall
+
+  * `forall| a. body` or `forall a.| body`
+
+  Either of these remove the `forall a.` prefix and leave just the body.
+
+  * `forall a|. type
+
+  If the type variable is a single character, replace the type variable with a TyPat.Hole,
+  otherwise remove the last character of the type variable.
+
+  * `forall a. type|`
+
+  If the body is not a type hole, this forwards the backspace action onto the
+  body, leaving the forall untouched. However, if the body is a type hole, this
+  removes the whole type, leaving only a type hole.
+
+## Type Vars
+
+  * `(type| a)` or `(type a)|`
```

Comment:
Remember that "(type ty)" allows ty to be any type expression, not just a type variable.

Reply:
Thanks! Yep addressed in new branch